docs/src/main/asciidoc/security-customization.adoc

```diff
@@ -620,6 +620,78 @@ are using an executor that is capable of propagating the identity (e.g. no `Comp
+* `io.quarkus.security.spi.runtime.AuthorizationSuccessEvent`
+* `io.quarkus.oidc.SecurityEvent`
+
+[[TIP]]
+For more information about security events specific to the Quarkus OpenID Connect extension, please see
+the xref:security-oidc-code-flow-authentication.adoc#listen-to-authentication-events[Listening to important authentication events]
+section of the OIDC code flow mechanism for protecting web applications guide.
+
+[source,java]
+----
+package org.acme.security;
+
+import io.quarkus.security.spi.runtime.AuthenticationFailureEvent;
+import io.quarkus.security.spi.runtime.AuthenticationSuccessEvent;
+import io.quarkus.security.spi.runtime.AuthorizationFailureEvent;
+import io.quarkus.security.spi.runtime.AuthorizationSuccessEvent;
+import io.quarkus.security.spi.runtime.SecurityEvent;
+import io.vertx.ext.web.RoutingContext;
+import jakarta.enterprise.event.Observes;
+import jakarta.enterprise.event.ObservesAsync;
+import org.jboss.logging.Logger;
+
+public class SecurityEventObserver {
+
+    private static final Logger LOG = Logger.getLogger(SecurityEventObserver.class.getName());
+
+    void observeAuthenticationSuccess(@ObservesAsync AuthenticationSuccessEvent event) {        <1>
+        LOG.debugf("User '%s' has authenticated successfully", event.getSecurityIdentity().getPrincipal().getName());
+    }
+
+    void observeAuthenticationFailure(@ObservesAsync AuthenticationFailureEvent event) {
+        RoutingContext routingContext = (RoutingContext) event.getEventProperties().get(RoutingContext.class.getName());
+        LOG.debugf("Authentication failed, request path: '%s'", routingContext.request().path());
+    }
+
+    void observeAuthorizationSuccess(@ObservesAsync AuthorizationSuccessEvent event) {
+        String principalName = getPrincipalName(event);
+        if (principalName != null) {
+            LOG.debugf("User '%s' has been authorized successfully", event.getSecurityIdentity().getPrincipal().getName());
+        }
+    }
+
+    void observeAuthorizationFailure(@Observes AuthorizationFailureEvent event) {
+        LOG.debugf(event.getAuthorizationFailure(), "User '%s' authorization failed", event.getSecurityIdentity().getPrincipal().getName());
+    }
+
+    private static String getPrincipalName(SecurityEvent event) {
+        if (event.getSecurityIdentity() != null) {                                              <2>
+            return event.getSecurityIdentity().getPrincipal().getName();
+        }
+        return null;
+    }
+
+}
+----
+<1> This observer consumes all the `AuthenticationSuccessEvent` events asynchronously, which means that HTTP request processing will continue regardless on the event processing.
+Depending on the application, that can be a lot of the `AuthenticationSuccessEvent` events.
+For that reason, asynchronous processing can have positive effect on performance.
+<2> PermitAll security check does not require authentication, therefore it can happen that SecurityIdentity will be null.
```

Comment:
```suggestion
<2> Common code for all supported security event types is possible because they all support `io.quarkus.security.spi.runtime.SecurityEvent` interface.
```

I got confused with a reference to `PermitAll`, it looks normal to protect against NPE without this clarification, so I propose instead to highlight the important point of being able to write a common code for all the event types